### PR TITLE
Run PyDREAM with python 3.8

### LINKED ISSUE #16: DreamPool implementation seems incompatible with Python3.8 multiprocess?
I have been having troble getting PyDream to work with Python 3.8 and I think I have tracked down the root of the problem: The `DreamPool`class  integrates with a multiprocess function that has changed in the Python3.8 version of multiprocess.

## What?
In Python3.8 when installing pyDream in a clean environment, a very simple test script kept outputting the error:
```
  File "/virtual_envs/py38_dream/lib/python3.8/site-packages/multiprocess/pool.py", line 212, in __init__
    self._repopulate_pool()
  File "/virtual_envs/py38_dream/lib/python3.8/site-packages/multiprocess/pool.py", line 303, in _repopulate_pool
    return self._repopulate_pool_static(self._ctx, self.Process,
  File "/virtual_envs/py38_dream/lib/python3.8/site-packages/multiprocess/pool.py", line 319, in _repopulate_pool_static
    w = Process(ctx, target=worker,
  File "/PyDREAM/pydream/Dream.py", line 996, in __init__
    mp.Process.__init__(self, group, target, name, args, kwargs)
  File "virtual_envs/py38_dream/lib/python3.8/site-packages/multiprocess/process.py", line 82, in __init__
    assert group is None, 'group argument must be None for now'
AssertionError: group argument must be None for now
```
When doing the exact same thing (environment, pip install ., try script) it works as expected.

## Why?
In previous implementations of multiprocess, the `Pool` class defined processes  with just args and kw args [see here](https://github.com/uqfoundation/multiprocess/blob/9d5e8155c26e4a4bcab6fd6699542fb17ba362fd/py3.6/multiprocess/pool.py#L150).
In the Python3.8 version `Pool` has an extra argument `ctx` [see here](https://github.com/uqfoundation/multiprocess/blob/9d5e8155c26e4a4bcab6fd6699542fb17ba362fd/py3.8/multiprocess/pool.py#L180).
In PyDream the `Process` member of `Pool` gets overwritten by the custom class [`NoDaemonProcess`](https://github.com/LoLab-VU/PyDREAM/blob/dd9cd08b2154ef5f45d7f0047817b327f64c2b0f/pydream/Dream.py#L1010). `NoDaemonProcess`does not expect to be passed the `ctx` variable, and interprets is as the `group` argument. `group` needs to be `None`, thus the error. 

## Ideas
In order to be compatible with multiprocess, I assume we need to make `NoDaemonProcess` depend on ctx (which is a `multiprocess.context.ForkContext` object, which incidentally hides another layer of indirection with another `Process` subclass). Unfortunately, I'm not really understanding what the `ctx` variable does.

I tried simply catching it with 
```
    def Process(self, ctx, *args, **kwds):
        return NoDaemonProcess(*args, **kwds)
```
and that seemed to work, I'm not entirely sure of whether there are unwanted consequences of this.

### PR DESCRIPTION
**Pull request by ortega2247**
multiprocessing has three different implementations of pool.Pool().Process() according to the python version (2.7, 3.4-3.7.0, 3.7.1) This PR copies the solution provided in the [nipype package](https://github.com/nipy/nipype/blob/master/nipype/pipeline/plugins/legacymultiproc.py). Also, I added a new argument that enables python 3 users to choose the multiprocessing context used to run a PyDREAM calibration.  Fixes #16

### CODE REVIEW

File: pydream/Dream.py [hunk 1]
```diff
@@ -55,11 +55,18 @@ class Dream():
         A model name to be used as a prefix when saving history and crossover value files.
     hardboundaries : bool
         Whether to relect point back into bounds of hard prior (i.e., if using a uniform prior, reflect points outside of boundaries back in, so you don't waste time looking at points with logpdf = -inf).
+    mp_context : a multiprocessing context or None. It will be used to launch the workers
```
Comment (alubbock) [inline pydream/Dream.py]: This should be in the format `mp_context: <type>` with the description below. Also would be helpful to add a comment about what None does, and when this option should be changed.

File: pydream/Dream.py [hunk 2]
```diff
@@ -990,22 +995,81 @@ def metrop_select(mr, q, q0):
         # Reject proposed value
         return q0
 
-        
-class NoDaemonProcess(mp.Process):
-    def __init__(self, group=None, target=None, name=None, args=(), kwargs={}):
-        mp.Process.__init__(self, group, target, name, args, kwargs)
```
Comment (alubbock) [inline pydream/Dream.py]: Add Attribution

File: pydream/Dream.py [hunk 3]
```diff
@@ -990,22 +995,81 @@ def metrop_select(mr, q, q0):
...
 
-    # make 'daemon' attribute always return False
-    def _get_daemon(self):
+# Pythons 2.7, 3.4-3.7.0, and 3.7.1 have three different implementations of
```
Comment (alubbock) [inline pydream/Dream.py]: Drop Python 2.7!

File: pydream/Dream.py [hunk 4]
```diff
@@ -990,22 +995,81 @@ def metrop_select(mr, q, q0):
...
 
-#A subclass of multiprocessing.pool.Pool that allows processes to launch child processes (this is necessary for Dream to use multi-try)
-#Taken from http://stackoverflow.com/questions/6974695/python-process-pool-non-daemonic
-class DreamPool(mp_pool.Pool):
-    def __init__(self, processes=None, initializer=None, initargs=None, maxtasksperchild=None):
-        mp_pool.Pool.__init__(self, processes, initializer, initargs, maxtasksperchild)
-    Process = NoDaemonProcess
 
+    class DreamPool(pool.Pool):
+        Process = NonDaemonProcess
```
Comment (alubbock) [inline pydream/Dream.py]: New line at end of file

File: pydream/core.py [hunk 5]
```diff
@@ -28,6 +29,7 @@ def run_dream(parameters, likelihood, nchains=5, niterations=50000, start=None,
         Whether to print verbose output (including acceptance or rejection of moves and the current acceptance rate).  Default: True
     tempering: Boolean, optional
         Whether to use parallel tempering for the DREAM chains.  Warning: this feature is untested.  Use at your own risk! Default: False
+    mp_context: multiprocessing context or None. It will be used to launch the workers
```
Comment (alubbock) [inline pydream/core.py]: Same issue as above with docstring

File: pydream/core.py [hunk 6]
```diff
@@ -51,11 +53,15 @@ def run_dream(parameters, likelihood, nchains=5, niterations=50000, start=None,
...
+                              history_file=kwargs['model_name'] + '_DREAM_chain_history.npy',
+                              crossover_file=kwargs['model_name'] + '_DREAM_chain_adapted_crossoverprob.npy',
+                              gamma_file=kwargs['model_name'] + '_DREAM_chain_adapted_gammalevelprob.npy',
+                              verbose=verbose, mp_context=mp_context, **kwargs)
     else:
-        step_instance = Dream(model=model, variables=parameters, verbose=verbose, **kwargs)
+        step_instance = Dream(model=model, variables=parameters, verbose=verbose, mp_context=mp_context, **kwargs)
 
-    pool = _setup_mp_dream_pool(nchains, niterations, step_instance, start_pt=start)
+    pool = _setup_mp_dream_pool(nchains, niterations, step_instance, start_pt=start, mp_context=mp_context)
```
Comment (alubbock) [inline pydream/core.py]: Add `try:` here

File: pydream/core.py [hunk 7]
```diff
@@ -72,7 +78,8 @@ def run_dream(parameters, likelihood, nchains=5, niterations=50000, start=None,
         returned_vals = pool.map(_sample_dream, args)
         sampled_params = [val[0] for val in returned_vals]
         log_ps = [val[1] for val in returned_vals]
```
Comment (alubbock) [inline pydream/core.py]: Add `finally:` here

**Comment by ortega2247** [comment]
Since the requested changes were minimal and the tests passed I'm going to go ahead and merge this PR

**Pull request merged**